### ISSUE TITLE
Azure provider: failure due concurrent access when requires exclusive access to Network Security Group

### ISSUE THREAD

**Issue by keymon** [hashicorp/terraform#3111]
If the provider tries to while creating Network Security Group rules in parallel it wil fail with: `Code: ConflictError, Message: Another operation that requires exclusive access to Network Security Group <group> is ongoing. Please try again later.`.

This happens just using `count` in the resource, [as here](https://github.com/alphagov/cf-terraform/commit/06188f903f6b1ae2a4363248c784ea1ad78dca51):

```
variable "office_cidrs" {
  description = "CSV of CIDR addresses for our office which will be trusted"
  default     = "80.194.77.90/32,80.194.77.100/32"
}
resource "azure_security_group" "bastion" {
    name = "bastion"
    location = "West Europe"
}
resource "azure_security_group_rule" "bastion_ssh_access" {
    count = "${length(split(",", "${var.office_cidrs}"))}"
    name = "bastion-ssh-access-rule-${count.index}"
    security_group_names = ["${azure_security_group.bastion.name}"]
    type = "Inbound"
    action = "Allow"
    priority = "20${count.index}"
    source_address_prefix = "${element(split(",", "${var.office_cidrs}"), count.index)}"
    source_port_range = "*"
    destination_address_prefix = "${azure_instance.bastion.vip_address}/32"
    destination_port_range = "22"
    protocol = "TCP"
}
```

The issue happens creating, updating and deleting.

Additionally, when updating or deleting, **terraform writes in the state** despite the actual object was NOT updated or deleted. `terraform refresh` does not fixes the state. 

Full error output:

```
$ terraform apply -var env=hectorjimazure 
azure_security_group.bastion: Refreshing state... (ID: hectorjimazure-bastion)
azure_hosted_service.bastion: Refreshing state... (ID: hectorjimazure-cf-bastion-service)
azure_storage_service.cf-storage: Refreshing state... (ID: hectorjimazurecfstorage)
azure_virtual_network.default: Refreshing state... (ID: hectorjimazure-default-network)
azure_instance.bastion: Refreshing state... (ID: hectorjimazure-cf-bastion)
azure_security_group_rule.bastion_block: Refreshing state... (ID: hectorjimazure-bastion-block-rule)
azure_security_group_rule.bastion_block: Modifying...
  priority: "60000" => "999"
azure_security_group_rule.bastion_ssh_access.1: Creating...
  action:                          "" => "Allow"
  destination_address_prefix:      "" => "10.0.0.4/32"
  destination_port_range:          "" => "22"
  name:                            "" => "hectorjimazure-bastion-ssh-access-rule-1"
  priority:                        "" => "201"
  protocol:                        "" => "TCP"
  security_group_names.#:          "" => "1"
  security_group_names.1368319355: "" => "hectorjimazure-bastion"
  source_address_prefix:           "" => "80.194.77.100/32"
  source_port_range:               "" => "*"
  type:                            "" => "Inbound"
azure_security_group_rule.bastion_ssh_access.0: Creating...
  action:                          "" => "Allow"
  destination_address_prefix:      "" => "10.0.0.4/32"
  destination_port_range:          "" => "22"
  name:                            "" => "hectorjimazure-bastion-ssh-access-rule-0"
  priority:                        "" => "200"
  protocol:                        "" => "TCP"
  security_group_names.#:          "" => "1"
  security_group_names.1368319355: "" => "hectorjimazure-bastion"
  source_address_prefix:           "" => "80.194.77.90/32"
  source_port_range:               "" => "*"
  type:                            "" => "Inbound"
azure_security_group_rule.bastion_ssh_access.0: Creation complete
azure_security_group_rule.bastion_block: Modifications complete
Error applying plan:

1 error(s) occurred:

* azure_security_group_rule.bastion_ssh_access.1: Error creating Azure network security group rule for security group "hectorjimazure-bastion": Error response from Azure. Code: ConflictError, Message: Another operation that requires exclusive access to Network Security Group hectorjimazure-bastion is ongoing. Please try again later.

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.

```

And updating:

```
$ terraform apply -var env=hectorjimazure 
azure_storage_service.cf-storage: Refreshing state... (ID: hectorjimazurecfstorage)
azure_hosted_service.bastion: Refreshing state... (ID: hectorjimazure-cf-bastion-service)
azure_security_group.bastion: Refreshing state... (ID: hectorjimazure-bastion)
azure_virtual_network.default: Refreshing state... (ID: hectorjimazure-default-network)
azure_instance.bastion: Refreshing state... (ID: hectorjimazure-cf-bastion)
azure_security_group_rule.bastion_block: Refreshing state... (ID: hectorjimazure-bastion-block-rule)
azure_security_group_rule.bastion_ssh_access.0: Refreshing state... (ID: hectorjimazure-bastion-ssh-access-rule-0)
azure_security_group_rule.bastion_ssh_access.1: Refreshing state... (ID: hectorjimazure-bastion-ssh-access-rule-1)
azure_security_group_rule.bastion_ssh_access.0: Modifying...
  destination_address_prefix: "10.0.0.4/32" => "40.113.120.178/32"
azure_security_group_rule.bastion_ssh_access.1: Modifying...
  destination_address_prefix: "10.0.0.4/32" => "40.113.120.178/32"
azure_security_group_rule.bastion_ssh_access.1: Modifications complete
Error applying plan:

1 error(s) occurred:

* azure_security_group_rule.bastion_ssh_access.0: Error updating Azure network security group rule for security group "hectorjimazure-bastion": Error response from Azure. Code: ConflictError, Message: Another operation that requires exclusive access to Network Security Group hectorjimazure-bastion is ongoing. Please try again later.

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.
```

**Comment by keymon** [comment]
To elaborate how the `terraform.state` gets out of sync with the real state of azure:

For instance, changing the nsg rules to point to the VIP instead of internal IPs, after the previous error we end in this state:

```
$ azure network nsg rule list hectorjimazure-bastion
info:    Executing command network nsg rule list
+ Looking up the network security group "hectorjimazure-bastion"               
data:    Name                                      Source address prefix  Source Port  Destination address prefix  Destination Port  Protocol  Type      Action  Priority
data:    ----------------------------------------  ---------------------  -----------  --------------------------  ----------------  --------  --------  ------  --------
data:    hectorjimazure-bastion-ssh-access-rule-0  80.194.77.90/32        *            10.0.0.4/32                 22                TCP       Inbound   Allow   200     
data:    hectorjimazure-bastion-ssh-access-rule-1  80.194.77.100/32       *            40.113.120.178/32           22                TCP       Inbound   Allow   201     
data:    hectorjimazure-bastion-block-rule         0.0.0.0/0              *            40.113.120.178/32           *                 TCP       Inbound   Deny    999     
data:    ALLOW VNET INBOUND                        VIRTUAL_NETWORK        *            VIRTUAL_NETWORK             *                 *         Inbound   Allow   65000   
data:    ALLOW VNET OUTBOUND                       VIRTUAL_NETWORK        *            VIRTUAL_NETWORK             *                 *         Outbound  Allow   65000   
data:    ALLOW INTERNET OUTBOUND                   *                      *            INTERNET                    *                 *         Outbound  Allow   65001   
data:    ALLOW AZURE LOAD BALANCER INBOUND         AZURE_LOADBALANCER     *            *                           *                 *         Inbound   Allow   65001   
data:    DENY ALL OUTBOUND                         *                      *            *                           *                 *         Outbound  Deny    65500   
data:    DENY ALL INBOUND                          *                      *            *                           *                 *         Inbound   Deny    65500   
info:    network nsg rule list command OK

```

But terraform.state says the desired state and does not apply any change. Terraform refresh did not solve the issue:

```
...
                "azure_security_group_rule.bastion_ssh_access.0": {
                    "type": "azure_security_group_rule",
                    "depends_on": [
                        "azure_instance.bastion",
                        "azure_security_group.bastion"
                    ],
                    "primary": {
                        "id": "hectorjimazure-bastion-ssh-access-rule-0",
                        "attributes": {
                            "action": "Allow",
                            "destination_address_prefix": "40.113.120.178/32",
                            "destination_port_range": "22",
...
```

```
keymon@keyair.local:~/gds/cf-terraform/azure$ terraform refresh -var env=hectorjimazure 
azure_storage_service.cf-storage: Refreshing state... (ID: hectorjimazurecfstorage)
azure_security_group.bastion: Refreshing state... (ID: hectorjimazure-bastion)
azure_hosted_service.bastion: Refreshing state... (ID: hectorjimazure-cf-bastion-service)
azure_virtual_network.default: Refreshing state... (ID: hectorjimazure-default-network)
azure_instance.bastion: Refreshing state... (ID: hectorjimazure-cf-bastion)
azure_security_group_rule.bastion_ssh_access.0: Refreshing state... (ID: hectorjimazure-bastion-ssh-access-rule-0)
azure_security_group_rule.bastion_block: Refreshing state... (ID: hectorjimazure-bastion-block-rule)
azure_security_group_rule.bastion_ssh_access.1: Refreshing state... (ID: hectorjimazure-bastion-ssh-access-rule-1)

Outputs:

  bastion-ips  = 10.0.0.4
  bastion-vips = 40.113.120.178
$ terraform apply -var env=hectorjimazure 
azure_security_group.bastion: Refreshing state... (ID: hectorjimazure-bastion)
azure_hosted_service.bastion: Refreshing state... (ID: hectorjimazure-cf-bastion-service)
azure_storage_service.cf-storage: Refreshing state... (ID: hectorjimazurecfstorage)
azure_virtual_network.default: Refreshing state... (ID: hectorjimazure-default-network)
azure_instance.bastion: Refreshing state... (ID: hectorjimazure-cf-bastion)
azure_security_group_rule.bastion_ssh_access.0: Refreshing state... (ID: hectorjimazure-bastion-ssh-access-rule-0)
azure_security_group_rule.bastion_block: Refreshing state... (ID: hectorjimazure-bastion-block-rule)
azure_security_group_rule.bastion_ssh_access.1: Refreshing state... (ID: hectorjimazure-bastion-ssh-access-rule-1)

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

Outputs:

  bastion-ips  = 10.0.0.4
  bastion-vips = 40.113.120.178
```

```

But changing the value manually, it force terraform to actually change the state.

**Comment by aznashwan** [comment]
@keymon: thanks a lot for the amazing bug report!

I have added a mutex in https://github.com/hashicorp/terraform/pull/3695 to guard against such situations.

As for the state failing to update; that is an unintentional byproduct of a design decision we made early on in Security Group Rule definitions. I have added a usage warning to the documentation [here](https://github.com/hashicorp/terraform/pull/3695/files#diff-037ccf6c8ce6a770d85ae6fa411bb707R14).
